CONTRIBUTING.md

```diff
@@ -0,0 +1,113 @@
+Note that some files (`Makefile`,`tests/*.go`,`tests/testdata/images/*.gz`) require all plugins to be rebuilt.
+Builds on the main branch don't use this optimization (we use concurrency controls to limit parallel builds and would need to determine what changed since the last published build).
+We may optimize this in the future.
+
+We use a combination of a custom command ([cmd/fetcher/main.go](https://github.com/bufbuild/plugins/blob/main/cmd/fetcher/main.go)) and Dependabot to keep dependencies up to date in the project.
+The `fetcher` command will use `source.yaml` files in each plugin to determine if new plugin versions are available.
+Dependabot is used to determine if base Docker images are up-to-date with bug/security fixes.
+
+### Caching
+
+We currently use [Buildkit registry caching](https://github.com/moby/buildkit#registry-push-image-and-cache-separately) to optimize build times for PR and main branch builds.
+Main branch builds publish Docker images to:
+
+* `ghcr.io/bufbuild/plugins-<org>-<name>:<version>` (Plugin image)
+* `ghcr.io/bufbuild/plugins-<org>-<name>:<version>-buildcache` (Registry cache with `mode=max`)
+
+These images are used only for caching - the authoritative images used for plugin execution are pushed to the BSR.
+Untagged versions of these cached images can be cleaned up at any time (only the latest tagged versions are used).
+
+During development, we've encountered a number of issues which lead to cache misses and in turn cause unnecessary revisions of plugins to be published to the BSR.
+Workarounds are in place for the majority of issues, however we should be careful to ensure build reproducibility.
+One critical change to the build images is the removal of `--build-arg` arguments in base images and the migration from `type=inline` to `type=registry` caching.
+See [PR #41](https://github.com/bufbuild/plugins/pull/41) and [PR #38](https://github.com/bufbuild/plugins/pull/38) for the gory details.
+
+## Updates
+
+### Creating a new plugin version
+
+If the `fetcher` command opens a PR for a new version of an existing plugin, there are currently some manual steps required before merging:
+
+1. Check out the PR locally (i.e. `gh pr checkout <num>`).
+2. Perform any manual updates to `package.json`, `go.mod`, or other files (if used by the plugin).
+   If NPM or Go dependencies change, ensure that `npm i` or `go mod tidy` are run to ensure that the `package-lock.json` or `go.sum` files are kept up to date.
+3. Double check version numbers in `plugin.yaml` or other files to ensure they are bumped.
+   Often times, there may be dependencies (plugin or runtime dependencies) which need to be updated.
+4. Run `make` and `make test` to build the plugin and run the unit tests.
```

Comment:
This is the part I've been struggling with lately. I'd love if we could avoid rebuilding the world when we want to work/commit a single plugin version.

Ideally we could target a specific versioned directory, which I think is possible with some of the env variables. If yes, should these instructions be updated?

Iirc the hesitation was that we open _multiple_ plugin versions updates per PR?

Reply:
If you've previously built everything it *should* only rebuild new/changed plugins on subsequent calls to make but I agree it would be helpful to target which plugins to build by name for example. I'll see what I can do here.